File: src/main/java/com/github/kuro46/embedscript/script/ScriptManager.java
package com.github.kuro46.embedscript.script;

import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

/**
 * Script manager<br>
 * This class is thread-safe
 */
public class ScriptManager {
    private final ConcurrentMap<ScriptPosition, List<Script>> scripts;
    private final Path path;

    public ScriptManager(ConcurrentMap<ScriptPosition, List<Script>> scripts, Path path) {
        this.scripts = scripts;
        this.path = path;
    }

    public static ScriptManager load(Path filePath) throws IOException {
        return new ScriptManager(new ConcurrentHashMap<>(ScriptSerializer.deserialize(filePath)), filePath);
    }

    public Path getPath() {
        return path;
    }

    private List<Script> getAndPutIfNeeded(ScriptPosition position) {
        return scripts.computeIfAbsent(position, ignore -> new ArrayList<>(1));
    }

    public boolean contains(ScriptPosition position) {
        return scripts.containsKey(position);
    }

    public List<Script> get(ScriptPosition position) {
        return scripts.getOrDefault(position, Collections.emptyList());
    }

    public void put(ScriptPosition position, Script script) {
        List<Script> scripts = getAndPutIfNeeded(position);
        scripts.add(script);
        ScriptSerializer.serializeLaterAsync(path, this.scripts);
    }

    public void putIfAbsent(ScriptPosition position, Script script) {
        if (scripts.containsKey(position)) {
            return;
        }
        List<Script> scripts = getAndPutIfNeeded(position);
        scripts.add(script);
        ScriptSerializer.serializeLaterAsync(path, this.scripts);
    }

    public List<Script> remove(ScriptPosition position) {
        List<Script> s = scripts.remove(position);
        ScriptSerializer.serializeLaterAsync(path, scripts);
        return s;
    }

    public Set<ScriptPosition> keySet() {
        return scripts.keySet();
    }

    public Set<Map.Entry<ScriptPosition, List<Script>>> entrySet() {
        return scripts.entrySet();
    }

    public void reload() throws IOException {
        Map<ScriptPosition, List<Script>> scripts = ScriptSerializer.deserialize(path);
        this.scripts.clear();
        this.scripts.putAll(scripts);
    }

    public void save() throws IOException {
        ScriptSerializer.serialize(path, scripts);
    }

    public void saveAsync() {
        ScriptSerializer.serializeLaterAsync(path, scripts);
    }

    public ConcurrentMap<ScriptPosition, List<Script>> getScripts() {
        return scripts;
    }
}

File: src/main/kotlin/com/github/kuro46/embedscript/script/ScriptManager.kt
package com.github.kuro46.embedscript.script

import java.nio.file.Path
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * Script manager<br></br>
 * This class is thread-safe
 *
 * @author shirokuro
 */
class ScriptManager(private val loader: Loader = NoOpLoader) {

    private val lock = ReentrantLock()
    @Volatile
    private var scripts: Map<ScriptPosition, List<Script>> = loader.initialValue()

    private fun setScripts(scripts: Map<ScriptPosition, List<Script>>) {
        this.scripts = scripts
    }

    fun getScripts(): Map<ScriptPosition, List<Script>> {
        return scripts
    }

    // =============
    // IO OPERATIONS
    // =============

    fun reload() {
        val loaded = loader.load()
        setScripts(loaded.scripts)
    }

    fun save() {
        loader.save(this)
    }

    fun saveAsync() {
        loader.saveAsync(this)
    }

    // ===============
    // READ OPERATIONS
    // ===============

    operator fun contains(position: ScriptPosition): Boolean {
        return scripts.containsKey(position)
    }

    operator fun get(position: ScriptPosition): List<Script>? {
        return scripts[position]
    }

    fun forEach(function: (ScriptPosition, List<Script>) -> Unit) {
        getScripts().forEach(function)
    }

    // ================
    // WRITE OPERATIONS
    // ================

    private fun shallowCopy(): MutableMap<ScriptPosition, List<Script>> {
        return HashMap(scripts)
    }

    fun remove(position: ScriptPosition): List<Script>? {
        return lock.withLock {
            val scripts = shallowCopy()
            val removed = scripts.remove(position)
            setScripts(scripts)
            loader.saveAsync(this, true)

            removed
        }
    }

    fun add(position: ScriptPosition, script: Script) {
        lock.withLock {
            val copied = shallowCopy()
            copied.compute(position) { _, current ->
                val addTo = current?.let { ArrayList(it) } ?: ArrayList(1)
                addTo.add(script)
                addTo
            }

            setScripts(copied)

            loader.saveAsync(this, true)
        }
    }

    fun addAll(position: ScriptPosition, scripts: List<Script>) {
        lock.withLock {
            val copied = shallowCopy()
            copied.compute(position) { _, current ->
                val addTo: MutableList<Script> = current?.let {
                    val copiedList: MutableList<Script> = ArrayList(it.size + scripts.size)
                    copiedList.addAll(it)

                    copiedList
                } ?: ArrayList(scripts.size)

                addTo.addAll(scripts)

                addTo
            }

            setScripts(copied)

            loader.saveAsync(this, true)
        }
    }

    interface Loader {
        fun save(scriptManager: ScriptManager, autoSave: Boolean = false)
        fun saveAsync(scriptManager: ScriptManager, autoSave: Boolean = false)
        fun load(): ScriptManager
        fun initialValue(): Map<ScriptPosition, List<Script>>
    }
}

object NoOpLoader : ScriptManager.Loader {
    override fun save(scriptManager: ScriptManager, autoSave: Boolean) {
        if (autoSave) return
        throw UnsupportedOperationException()
    }

    override fun saveAsync(scriptManager: ScriptManager, autoSave: Boolean) {
        if (autoSave) return
        throw UnsupportedOperationException()
    }

    override fun load(): ScriptManager {
        throw UnsupportedOperationException()
    }

    override fun initialValue(): Map<ScriptPosition, List<Script>> {
        return emptyMap()
    }
}

class JsonLoader(val path: Path) : ScriptManager.Loader {
    override fun save(scriptManager: ScriptManager, autoSave: Boolean) {
        ScriptSerializer.serialize(path, scriptManager)
    }

    override fun saveAsync(scriptManager: ScriptManager, autoSave: Boolean) {
        ScriptSerializer.serializeLaterAsync(path, scriptManager)
    }

    override fun load(): ScriptManager {
        return ScriptSerializer.deserialize(path)
    }

    override fun initialValue(): Map<ScriptPosition, List<Script>> {
        return ScriptSerializer.deserialize(path).getScripts()
    }
}
